crates/next-core/src/util.rs

```diff
@@ -66,3 +73,155 @@ pub async fn foreign_code_context_condition(next_config: NextConfigVc) -> Result
     };
     Ok(result)
 }
+
+#[derive(Default, PartialEq, Eq, Clone, Copy, Debug, TraceRawVcs, Serialize, Deserialize)]
+pub enum NextRuntime {
+    #[default]
+    NodeJs,
+    Edge,
+}
+
+#[turbo_tasks::value]
+#[derive(Default)]
+pub struct NextSourceConfig {
+    pub runtime: NextRuntime,
+}
+
+/// An issue that occurred while resolving the React Refresh runtime module.
+#[turbo_tasks::value(shared)]
+pub struct NextSourceConfigParsingIssue {
+    path: FileSystemPathVc,
+    detail: StringVc,
+}
+
+#[turbo_tasks::value_impl]
+impl Issue for NextSourceConfigParsingIssue {
+    #[turbo_tasks::function]
+    fn severity(&self) -> IssueSeverityVc {
+        IssueSeverity::Warning.into()
+    }
+
+    #[turbo_tasks::function]
+    fn title(&self) -> StringVc {
+        StringVc::cell("Unable to parse config export in source file".to_string())
+    }
+
+    #[turbo_tasks::function]
+    fn category(&self) -> StringVc {
+        StringVc::cell("parsing".to_string())
+    }
+
+    #[turbo_tasks::function]
+    fn context(&self) -> FileSystemPathVc {
+        self.path
+    }
+
+    #[turbo_tasks::function]
+    fn description(&self) -> StringVc {
+        StringVc::cell(
+            "The exported configuration object in a source file need to have a very specific \
+             format from which some properties can be statically parsed at compiled-time."
+                .to_string(),
+        )
+    }
+
+    #[turbo_tasks::function]
+    fn detail(&self) -> StringVc {
+        self.detail
+    }
+}
+
+#[turbo_tasks::function]
+pub async fn parse_config_from_source(module_asset: AssetVc) -> Result<NextSourceConfigVc> {
+    if let Some(ecmascript_asset) = EcmascriptModuleAssetVc::resolve_from(module_asset).await? {
+        if let ParseResult::Ok { program, .. } = &*ecmascript_asset.parse().await? {
+            if let Program::Module(module) = &program {
+                for item in &module.body {
+                    if let Some(decl) = item
+                        .as_module_decl()
+                        .and_then(|mod_decl| mod_decl.as_export_decl())
+                        .and_then(|export_decl| export_decl.decl.as_var())
+                    {
+                        for decl in &decl.decls {
```

Comment:
I had commented before in slack that we could write a generic AST evaluator, and this might allow us to parse the `next.config.js` file without actually spinning up node. We have a similar desire here, and we could make this more generic by using a function that takes the export were interested in, and returning a `Result<JsonValue>` of the statically analyzed value.

Reply:
Technically we have such an evaluator for the static analysis. Maybe we can use that here instead.

Reply:
did that